src/blocktridiag.jl

```diff
@@ -0,0 +1,243 @@
+#=======================
+ BlockTridiagMatrix
+=======================#
+
+"""
+    BlockTridiagMatrix{T, R <: AbstractMatrix{T}} <: AbstractBlockArray{T, N}
+
+A `BlockTridiagMatrix` is a block tridiagonal matrix where each block is stored contiguously.
+This means that insertions and retrieval of blocks
+can be very fast and non allocating since no copying of data is needed.
+
+In the type definition, `R` defines the array type that each block has,
+for example `Matrix{Float64}.
+"""
+struct BlockTridiagMatrix{T, R <: AbstractMatrix{T}} <: AbstractBlockMatrix{T}
+    diagl::Vector{R}
+    lower::Vector{R}
+    upper::Vector{R}
+    block_sizes::BlockSizes{2}
+end
+
+# Auxilary outer constructors
+function BlockTridiagMatrix{T, R <: AbstractArray{T}
+                           }(diagl::Vector{R},
+                             lower::Vector{R},
+                             upper::Vector{R},
+                             block_sizes::BlockSizes{2})
+   return BlockTridiagMatrix{T, R}(diagl, lower, upper, block_sizes)
+end
+
+function BlockTridiagMatrix{T, R <: AbstractArray{T}
+                           }(diagl::Vector{R},
+                             lower::Vector{R},
+                             upper::Vector{R},
+                             block_sizes::Vararg{Vector{Int}, 2})
+    return BlockTridiagMatrix{T, R}(diagl, lower, upper,
+                                    BlockSizes(block_sizes...))
+end
+
+
+################
+# Constructors #
+################
+
+"""
+Constructs a `BlockTridiagMatrix` with uninitialized blocks from a block type `R`
+with sizes defind by `block_sizes`.
+
+```jldoctest
+julia> BlockTridiagMatrix(Matrix{Float64}, [1,3], [2,2])
+2×2-blocked 4×4 BlockArrays.BlockTridiagMatrix{Float64,2,Array{Float64,2}}:
+ #undef  │  #undef  #undef  #undef  │
+ --------┼--------------------------┼
+ #undef  │  #undef  #undef  #undef  │
+ #undef  │  #undef  #undef  #undef  │
+ --------┼--------------------------┼
+ #undef  │  #undef  #undef  #undef  │
+```
+"""
+@inline function BlockTridiagMatrix{T, R <: AbstractMatrix{T}
+                                   }(::Type{R},
+                                     block_sizes::Vararg{Vector{Int}, 2})
+    BlockTridiagMatrix(R, BlockSizes(block_sizes...))
+end
+
+function BlockTridiagMatrix{T, R <: AbstractMatrix{T}}(::Type{R}, block_sizes::BlockSizes{2})
+    n_blocks = nblocks(block_sizes)
+    n_blocks[1] == n_blocks[2] || throw("expect same number of blocks in both dimensions")
```

Comment:
This should probably throw an `ArgumentError`. I also prefer to write erros in "must"-form e.g `the number of blocks must be the same in both dimensions`.